crates/bevy_asset/macros/src/lib.rs

```diff
@@ -41,33 +41,101 @@ fn derive_dependency_visitor_internal(
     ast: &DeriveInput,
     bevy_asset_path: &Path,
 ) -> Result<proc_macro2::TokenStream, syn::Error> {
-    let mut field_visitors = Vec::new();
-    if let Data::Struct(data_struct) = &ast.data {
-        for field in &data_struct.fields {
-            if field
-                .attrs
-                .iter()
-                .any(|a| a.path().is_ident(DEPENDENCY_ATTRIBUTE))
-            {
-                if let Some(field_ident) = &field.ident {
-                    field_visitors.push(quote! {
-                        #bevy_asset_path::VisitAssetDependencies::visit_dependencies(&self.#field_ident, visit);
-                    });
+    let struct_name = &ast.ident;
+    let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
+
+    let body = match &ast.data {
+        Data::Struct(data_struct) => {
+            let mut field_visitors = Vec::new();
+
+            for field in &data_struct.fields {
+                if field
+                    .attrs
+                    .iter()
+                    .any(|a| a.path().is_ident(DEPENDENCY_ATTRIBUTE))
+                {
+                    if let Some(field_ident) = &field.ident {
+                        field_visitors.push(quote! {
+                            #bevy_asset_path::VisitAssetDependencies::visit_dependencies(&self.#field_ident, visit);
+                        });
+                    }
                 }
             }
+
+            quote! {
+                #(#field_visitors)*
+            }
         }
-    } else {
-        return Err(syn::Error::new(
-            Span::call_site().into(),
-            "Asset derive currently only works on structs",
-        ));
-    }
+        Data::Enum(data_enum) => {
+            let mut any_case_required = false;
+            let cases = data_enum.variants.iter().map(|variant| {
+                let mut any_member_omitted = false;
+                let mut field_names = Vec::new();
+                let mut field_visitors = Vec::new();
 
-    let struct_name = &ast.ident;
-    let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
+                for (i, field) in variant.fields.iter().enumerate() {
```

Comment:
Here as well, I don't think collecting is necessary.

Reply:
i'm building 2 separate vectors here, i think i need to collect but happy to be proved wrong